Fix syncing of local tokens with remote payment methods

a-danae opened the pull request:
Fixes #2896 

## Changes proposed in this Pull Request:

- Reduce the amount of iteration when syncing local tokens with the payment methods in Stripe. We had four loops, two of them nested within each other. Now we're using two, not nested.
- (Hopefully) Simplify the logic behind the syncing process.

## Testing instructions

**Pulling remote payment methods**
1. Log in as a shopper
2. Go to My Account > Payment methods
3. Take note of the payment methods displayed
<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/071a3d34-8acc-46e3-b40f-1ef2085545d3">

4. On the Stripe dashboard, go to the page for the customer associated with this shopper. It's in `https://dashboard.stripe.com/test/customers/cus_xxxx`
5. Under Payment Methods, click on Add > Add card
6. Enter a test card, like `4000056655665556`
<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/9c5b04b2-6bc6-415c-9c8f-d056edc259fb">

7. Skip the stored transient. Either remove the transient in wp_options `LIKE '%stripe_payment_methods_%'`, or assign `false` to [this variable](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/fix/tokens-syncing/includes/class-wc-stripe-customer.php#L451)
8. As the shopper, reload the My Account > Payment methods page
9. Confirm that the card you just added to Stripe is listed
<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/af47ec9f-d897-4e04-8919-3d10f8d63323">

**Deleting local payment methods when removed remotely**
1. Log in as a shopper
2. Go to My Account > Payment methods
3. Take note of the payment methods displayed there
<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/7384488e-07d4-4a00-b3ac-eba6dbea900c">

4. On the Stripe dashboard, go to the page for the customer associated with this shopper. It's in `https://dashboard.stripe.com/test/customers/cus_xxxx`
5. Under Payment Methods, delete one of the payment methods
<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/dbd91cfc-003f-492b-a65e-4d9db0de7f1b">

6. Skip the stored transient. Either remove the transient in wp_options `LIKE '%stripe_payment_methods_%'`, or assign `false` to [this variable](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/fix/tokens-syncing/includes/class-wc-stripe-customer.php#L451)
7. As the shopper, reload the My Account > Payment methods page
8. Confirm that the payment method you deleted in Stripe is no longer listed
<img width="500" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/56d5cf11-639f-42cb-89aa-77926da21d9c">


**Regression tests**

Listing of payment methods from other extensions
1. Enable WooPayments
2. Save some card and SEPA payment methods
3. Visit the My Account > Payment methods page
4. Confirm all the tokens are listed as expected: The tokens are listed, and they don't get duplicated or deleted

Listing saved APMs
1. Purchase a subscription using iDEAL, SEPA, or Bancontact
2. Go to the My Account > Payment Methods page
3. Confirm all the tokens are listed as expected


Disabling specific payment method types with saved payment methods
1. As a shopper, save a card and a SEPA payment method
2. Go to the My Account > Payment Methods page and note the listed payment methods
3. As a merchant, go to the Stripe settings page in the WP dashboard > Payment methods tab
4. Disable SEPA
5. As a shopper, go to the My Account > Payment Methods page
6. Confirm no SEPA payment methods are listed

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

a-danae committed:
Refactor deleting local payment methods that don't exist in Stripe

a-danae committed:
Create tokens for payment methods missing locally

a-danae committed:
Handle exceptions when creating and deleting payment methods

a-danae committed:
Add missing variable for the payment method type

a-danae committed:
Bail out early when the token doesn't belong to our reusable payment gateways

Mayisha commented on includes/class-wc-stripe-payment-tokens.php:
```diff
@@ -250,65 +250,99 @@ public function woocommerce_get_customer_upe_payment_tokens( $tokens, $user_id,
-		$gateway     = WC_Stripe::get_instance()->get_main_stripe_gateway();
-		$upe_gateway = null;
+		try {
+			$gateway  = new WC_Stripe_UPE_Payment_Gateway();
```
Could there be any side effect of instantiating this class here? i.e. recreating or firing subscription hooks, etc?

a-danae replied:
Great insight. This actually was a problem in a previous PR. Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2897/commits/2aada620fad04fe3d4d1c4cbcd0155df4850e4f3

Mayisha commented on includes/class-wc-stripe-payment-tokens.php:
```diff
@@ -250,65 +250,99 @@ public function woocommerce_get_customer_upe_payment_tokens( $tokens, $user_id,
 
-			foreach ( $payment_methods as $payment_method ) {
-				if ( ! isset( $current_tokens[ $payment_method->id ] ) ) {
-					$payment_method_type = $this->get_original_payment_method_type( $payment_method );
```
As this `get_original_payment_method_type` is not used anymore, it can be removed I think.

a-danae replied:
Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2897/commits/207d5673c6959b91dc5541b04131cdb15d1c0159

Mayisha reviewed (commented):
I had already half reviewed this PR last Friday so decided to finish my review. I have tested the scenarios mentioned in the description and got the following outcomes.
- Pulling remote payment methods :heavy_check_mark: 
- Deleting local payment methods when removed remotely :heavy_check_mark: 
- Regression tests (tested with WooPayments by saving a card) :heavy_check_mark:
- Listing saved APMs (tested by purchasing a subscription using SEPA) :heavy_check_mark: 
- Disabling specific payment method types with saved payment methods (for iDeal and Bancontact) :heavy_check_mark: 

I found 2 problems.
- When sepa is enabled, I see it on the checkout page and even save the method while purchasing a subscription but I don't see it on `My Account` page. 

![no sepa](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/34f25db1-16d3-4b2c-a6a7-e59b1ff0fc4a)

-  I can not add a payment method from `My Account` page. I get the following error
```
Notice: Undefined index: return_url in /var/www/html/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-intent-controller.php on line 959 {"success":true,"data":{"status":"succeeded","id":"seti_1OkQJ2CBKj0yxC0ZfaFbFojV","client_secret":"seti_1OkQJ2CBKj0yxC0ZfaFbFojV_secret_PZZRODQG5VdX2jq6fUlByCTtxWcipL0"}}
```
The issues seem unrelated to this PR. Let me know your thoughts.

**Note**: This one is not important but thought I would add it discovered during testing. Adding/deleting a payment method works find both ways to/from Stripe Dashboard and My Account page. When I set a default method on My account page it's reflected on the Stripe Dashboard. But setting default from Stripe does not reflect on the My account page. It's a unlikely scenario though, as a merchant won't be setting a default for a shopper.

a-danae committed:
Retrieve the main gateway instance through WC_Stripe::get_main_stripe_gateway

Using this instead of creating a new instance of WC_Stripe_UPE_Payment_Gateway to avoid potential side-effects.

a-danae committed:
Remove unused WC_Stripe_Payment_Tokens::get_original_payment_method_type method

a-danae committed:
Merge branch 'add/deferred-intent' into fix/tokens-syncing

a-danae commented:
Thanks for the thorough review, @Mayisha !

> I can not add a payment method from My Account page. I get the following error

There's a GH issue for this over here https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2907

> When sepa is enabled, I see it on the checkout page and even save the method while purchasing a subscription but I don't see it on My Account page.

Good catch. I can replicate in `add/deferred-intent`, so I'd say we check this in a separate issue. ~Creating it...~ GH issue created https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2922

> This one is not important but thought I would add it discovered during testing. Adding/deleting a payment method works find both ways to/from Stripe Dashboard and My Account page. When I set a default method on My account page it's reflected on the Stripe Dashboard. But setting default from Stripe does not reflect on the My account page. It's a unlikely scenario though, as a merchant won't be setting a default for a shopper.

I can replicate this in `develop`. I also don't see any code intended to keep this option in sync. 
But it might not be an issue. Do you see any potential problems with keeping it as is? We could create a GH issue to spike whether we'd like to change this behavior.

Mayisha reviewed (approved):
Thanks Danae for the fix. Addressing the two issues with their separate tickets sounds good. 

> Do you see any potential problems with keeping it as is?

Also I don't see any problems here as it is not anything that significaant.
